### PR TITLE
[diffusion, rollout, reward, data, recipe] feat: add FLUX.1-dev DanceGRPO training

### PR DESCRIPTION
**Pull request by RhoHwai**
### What does this PR do?

Add an end-to-end, full-parameter FLUX.1-dev DanceGRPO training path to verl-omni, using vLLM-Omni for rollout, the existing diffusion trainer for Actor updates, and HPSv2.1 as the default reward.

The implementation:

- adds a FLUX.1-dev rollout adapter that preserves separate CLIP and T5 prompt token IDs, performs FLUX-native prompt encoding, applies the shifted DanceSDE schedule, and returns explicit current/next latent pairs with rollout log-probabilities;
- adds a diffusers-based Actor adapter that reconstructs the same FLUX inputs and transition schedule for DanceGRPO policy updates;
- reuses the existing DanceGRPO trainer, diffusion configuration contracts, scheduler utilities, and rollout/Actor interfaces instead of adding FLUX-specific branches to `ray_diffusion_trainer` or introducing new trainer configuration fields;
- adds an HPSv2.1 custom reward scorer with lazy per-worker loading and CPU inference by default;
- adds explicit train/test prompt conversion to the standard chat-style parquet schema;
- adds an Ascend NPU full-parameter launcher and documentation for model, data, and HPSv2 checkpoint paths;
- adds CPU tests for transition selection, SDE windows, token handling, rollout/Actor schedule alignment, Actor guidance defaults, and HPSv2 image conversion.

The default recipe uses 720×720 images, 16 denoising steps, rollout guidance `3.5`, Actor guidance `1.0`, scheduler shift `3.0`, DanceSDE noise `0.3`, and a random 60% transition subset after dropping the final transition.

This PR is intentionally limited to FLUX.1-dev full-parameter DanceGRPO training with HPSv2.1. HPSv3 and a FLUX LoRA recipe are out of scope. The provided launcher is Ascend NPU-only.

### Relationship to existing work and non-duplication

Related issue/PR: None.

### Test

Pre-commit checks for the PR-owned files and generated trainer configuration:

```bash
PR_TITLE='[diffusion, rollout, reward, data, recipe] feat: add FLUX.1-dev DanceGRPO training' \
python tests/special_sanity/check_pr_title.py

pre-commit run ruff --files <changed-files>
pre-commit run ruff-format --files <changed-files>
pre-commit run mypy --files <changed-files>
pre-commit run autogen-trainer-cfg --all-files

bash -n examples/dancegrpo_trainer/flux1/run_flux1_dev_dancegrpo_fullparam.sh

python -m compileall -q \
  examples/dancegrpo_trainer/flux1/dataprocess/prepare_prompts.py \
  verl_omni/pipelines/flux_dance_grpo \
  verl_omni/utils/reward_score/hpsv2_reward.py

git diff --check
```

Result: all commands above passed. The proposed title passed the repository title validator.

CPU tests added by this PR:

```bash
python -m pytest -q \
  tests/pipelines/test_flux_dance_grpo_on_cpu.py \
  tests/utils/reward_score/test_hpsv2_reward_on_cpu.py
```

These tests cover aligned transition gathering with `drop_last` and fractional sampling, SDE window normalization/sampling, token padding and required CLIP/T5 `extra_tokenizers` data, exact rollout/Actor schedule equality for 20 denoising steps, the Actor guidance default, and supported/invalid HPSv2 image inputs. The full CPU test suite completed successfully: `1128 passed, 0 failed` in `49.39s` with exit code `0`.

The example also documents a complementary real-weight, one-step end-to-end validation command:

```bash
TOTAL_TRAINING_STEPS=1 \
EXPERIMENT_NAME=flux1_dev_fullparam_hpsv2_smoke \
bash examples/dancegrpo_trainer/flux1/run_flux1_dev_dancegrpo_fullparam.sh
```

A successful smoke run reaches `training/global_step:1` and reports `critic/hpsv2_raw/mean`, `rollout_corr/logprob_abs_diff_mean`, `actor/loss`, and `timing_s/step`.

Real-weight Ascend validation:

- Eight Ascend Atlas A3 NPUs, 720×720 output, 16 denoising steps, full-parameter Actor training, and HPSv2.1 reward.
- `open_clip_torch==3.3.0` successfully loaded the OpenCLIP and HPSv2.1 checkpoints, and the reward workers produced HPSv2 scores during training.

#### Training results: first 100 comparison steps

The table reports arithmetic means for each displayed interval. Reward is the mean raw HPSv2 score per training step. The vLLM 0.24.0 workload normalization is explained below the table.

| Steps | MindSpeed-MM reward | MindSpeed-MM time (s/step) | verl-omni / vLLM 0.24.0 reward | verl-omni / vLLM 0.24.0 normalized time (s/step) | verl-omni / vLLM 0.27.1.dev0 reward | verl-omni / vLLM 0.27.1.dev0 time (s/step) | verl-omni / vLLM 0.28.0 reward | verl-omni / vLLM 0.28.0 time (s/step) |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1–10 | 0.304355 | 397.30 | 0.309064 | 295.94 | 0.310482 | 321.05 | Testing | Testing |
| 11–20 | 0.301006 | 404.34 | 0.311398 | 295.55 | 0.310261 | 320.07 | Testing | Testing |
| 21–30 | 0.317296 | 410.92 | 0.315434 | 297.71 | 0.312007 | 318.30 | Testing | Testing |
| 31–40 | 0.320794 | 292.78 | 0.314099 | 299.57 | 0.308585 | 319.68 | Testing | Testing |
| 41–50 | 0.321875 | 289.46 | 0.316091 | 293.72 | 0.311293 | 322.52 | Testing | Testing |
| 51–60 | 0.320441 | 279.06 | 0.323983 | 294.91 | 0.317975 | 323.81 | Testing | Testing |
| 61–70 | 0.325590 | 277.56 | 0.336245 | 295.83 | 0.327611 | 321.19 | Testing | Testing |
| 71–80 | 0.330947 | 277.46 | 0.326705 | 293.39 | 0.317322 | 320.30 | Testing | Testing |
| 81–90 | 0.329360 | 277.57 | 0.331612 | 293.25 | 0.330849 | 321.25 | Testing | Testing |
| 91–100 | 0.338033 | 277.77 | 0.333782 | 291.15 | 0.329052 | 321.01 | Testing | Testing |

Metric sources and runtime versions:

- MindSpeed-MM: `gathered_hps_reward_mean` and tqdm `step_time` from the original training log.
- verl-omni / vLLM 0.24.0: `critic/hpsv2_raw/mean` and `timing_s/step`; the source tree pins `vllm==0.24.0`, `vllm-omni==0.24.0`, and vLLM-Omni commit `fe478a95ab39e80085da06e3401c3ed471898a57`.
- verl-omni / vLLM 0.27.1.dev0: `critic/hpsv2_raw/mean` and `timing_s/step`; the log reports vLLM commit `4bdc8a788` and vLLM-Omni commit `444485650`.
- verl-omni / vLLM 0.28.0: the current dependency set pins `vllm==0.28.0`, `vllm-omni==0.28.0rc1`, and vLLM-Omni commit `ded8934626aaad1a3e816c3a1d9d742efc012d93`; testing is in progress.

flux1_reward_ma5_first120

Over the equivalent first 100 comparison steps, the observed mean HPSv2 rewards are `0.321841` for vLLM 0.24.0, `0.320970` for MindSpeed-MM, and `0.317544` for vLLM 0.27.1.dev0. The workload-normalized mean step time is `295.10` seconds for vLLM 0.24.0 and `320.92` seconds for vLLM 0.27.1.dev0, making the observed vLLM 0.24.0 result approximately 8.0% lower.

### API and Usage Example

Install the validated HPSv2 dependency on top of the standard verl-omni environment:

```bash
pip install open_clip_torch==3.3.0
```

Prepare explicit train and validation prompt files. The default output is `~/data/hpsv2`, which matches `$WORKSPACE/data/hpsv2` when `WORKSPACE` uses its default value of `$HOME`:

```bash
python3 examples/dancegrpo_trainer/flux1/dataprocess/prepare_prompts.py \
  --train-path /path/to/train.txt \
  --test-path /path/to/test.txt \
  --output-dir "${WORKSPACE:-$HOME}/data/hpsv2"
```

Launch training from the repository root with explicit model and reward-checkpoint paths:

```bash
export WORKSPACE=/path/to/workspace
export MODEL_NAME=/path/to/FLUX.1-dev
export HPSV2_PRETRAINED_PATH=/path/to/open_clip_pytorch_model.bin
export CUSTOM_REWARD_MODEL_PATH=/path/to/HPS_v2.1_compressed.pt

bash examples/dancegrpo_trainer/flux1/run_flux1_dev_dancegrpo_fullparam.sh
```

The launcher reads `$WORKSPACE/data/hpsv2/train.parquet` and `$WORKSPACE/data/hpsv2/test.parquet` by default. `TRAIN_FILES_PATH` and `VAL_FILES_PATH` can override those locations. Missing model or HPSv2 checkpoint variables fail immediately instead of falling back to machine-specific paths.

HPSv2 scoring defaults to two CPU-resident reward workers. Set `REWARD_DEVICE` explicitly only when a different scoring device is intended.

### Design & Code Changes

1. **Reuse the existing trainer and configuration contracts:** FLUX registers through the same diffusion model/rollout discovery path used by existing models. This PR does not modify `ray_diffusion_trainer` or add FLUX-only trainer config fields.
2. **FLUX-native conditioning:** transport separate CLIP and T5 token IDs and reuse the official FLUX prompt-embedding, packed-latent, rotary-position, and VAE conventions from diffusers and pinned vLLM-Omni.
3. **Aligned DanceGRPO rollout:** extend the vLLM-Omni FLUX pipeline with shifted sigmas, per-prompt DanceSDE windows, explicit transition pairing, log-probability capture, and deterministic initial-noise grouping.
4. **Aligned Actor replay:** rebuild the same FLUX transformer inputs on the training side and evaluate the stored transition under the Actor policy without a negative-prompt CFG branch. The Actor guidance default is `1.0`, matching the recipe's Actor/rollout split (`1.0`/`3.5`).
5. **Shared schedule behavior:** reuse Wan2.2's existing `sd3_time_shift` utility. Both rollout and Actor construct the base schedule with NumPy `linspace`, then apply the same float32 shift, so exact timestep lookup remains stable for non-dyadic step counts such as 20.
6. **HPSv2.1 reward integration:** lazily load one OpenCLIP scorer per reward worker, validate checkpoint paths, normalize supported image representations, and return both the standard score and raw HPSv2 metric.
7. **Reproducible recipe and data boundary:** keep train/test prompts explicit, write the existing parquet schema, require local model/checkpoint paths, and expose only operational overrides through environment variables.
8. **Fail-fast optional rollout import:** when `vllm-omni` is unavailable, export a raise-on-use placeholder instead of silently leaking `None` through `__all__`.

This change is larger than the preferred 500-line guideline because rollout trajectory capture, Actor replay, reward scoring, pipeline registration, prompt preparation, the launcher, and their CPU tests are required together to form a runnable and reviewable training path. Splitting any of these pieces into an independently merged PR would leave incomplete FLUX support that cannot be validated end to end. The implementation minimizes shared-framework changes: ten of the eleven changed files are feature-specific additions; the only shared-file change is a three-line pipeline export.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Synchronize the branch with upstream `main` at `03655dc`; merge commit `ee61dbd` completed without conflicts.
- [x] Run Ruff, Ruff format, mypy, generated-config verification, Shell syntax, Python compilation, and diff checks for the PR-owned changes.
- [x] Add the FLUX recipe documentation without changing the existing Wan2.2 DanceGRPO README.
- [x] Add CPU tests for the new FLUX DanceGRPO helpers/contracts and HPSv2 image conversion.
- [x] Run the full CPU test suite in the project environment: `1128 passed, 0 failed` in `49.39s` with exit code `0`.
- [x] Document the one-step real-weight validation command and expected metrics in the FLUX example.
- [ ] Complete and report current-pin vLLM 0.28.0 real-weight Ascend validation.
- [x] Record the Ascend Atlas A3 hardware used for the experiments.
- [x] Attach any test evidence/screenshots required for review.

### AI assistance and accountability

AI assistance (OpenAI Codex) was used to inspect, revise, validate, and document this change. Before requesting review, the human submitter will review every changed line, rerun the relevant checks in the project environment, and be prepared to explain and maintain the implementation.

### CODE REVIEW

**Comment by RhoHwai** [comment]
@zhtmike Could you please review this PR?

**Comment by zhtmike** [comment]
thanks for the detailed writeup and the MindSpeed-MM comparison table.

Verdict: NEEDS CHANGES — add the CPU tests and fix the two silent-trap spots below; the adapter design itself is fine (reuses the trainer, scheduler, and rollout contracts cleanly).

1. whole PR ships zero tests. Every existing adapter and scorer has a CPU test (`tests/pipelines/test_wan22_dance_grpo_on_cpu.py`, `tests/utils/reward_score/test_pickscore_reward_on_cpu.py`, ...), and your `flux_dance_grpo/__init__.py` guard says it exists "to permit CPU-only unit tests" — so ship them: `select_dance_grpo_transitions` (drop_last/fraction/gather alignment), `_sample_sde_windows` + `_normalize_window_range`, `_pad_token_ids`, `_to_rgb_image`, and the `extra_tokenizers` error paths. The PR-body argument against a "registry-only assertion" answers a strawman — the real-weight smoke run and the CPU tests are complements, not alternatives.

2. `vllm_omni_rollout_adapter.py:231` builds sigmas with `torch.linspace` but the actor (`diffusers_training_adapter.py:113`) rebuilds them with `np.linspace`; the actor then matches stored rollout timesteps against its own schedule via exact equality inside `index_for_timestep`, so a 1-ULP drift → IndexError mid-training. 16 steps happens to be dyadic-exact; 20 steps likely isn't. Use `np.linspace` on both sides, like the wan22 pair does.

3. `diffusers_training_adapter.py:213-214` — actor-side `guidance_scale` falls back to 3.5, but this PR's own design is actor 1.0 / rollout 3.5 (README says so). Someone who forgets `model.pipeline.guidance_scale` silently replays with distilled-guidance 3.5. Default to 1.0 here or require it.

4. `flux_dance_grpo/__init__.py:24` — exporting `FluxDanceGRPOPipelineWithLogProb = None` fails silently; wan22 substitutes a raise-on-use placeholder. Consider the same so a missing vllm-omni blows up at use instead of leaking None through `__all__`.

AI assistance (ZCode) was used for this review.

**Comment by zhtmike** [comment]
we need a reward_curve.png at the PR description.

**Comment by RhoHwai** [comment]
@zhtmike   Thank you for the detailed review. I have addressed all four requested changes:

1. Added CPU coverage in `tests/pipelines/test_flux_dance_grpo_on_cpu.py` and `tests/utils/reward_score/test_hpsv2_reward_on_cpu.py` for:
   - `select_dance_grpo_transitions`, including `drop_last`, fractional selection, per-row generators, and gathered-field alignment;
   - `_normalize_window_range` and `_sample_sde_windows`;
   - `_pad_token_ids` and the required CLIP/T5 `extra_prompt_ids` error paths;
   - `_to_rgb_image` conversion and invalid-input paths;
   - rollout/Actor sigma equality with 20 denoising steps; and
   - the Actor-side default guidance value.
2. Changed the rollout schedule construction to the same NumPy `linspace` -> float32 path used by the Actor. The new 20-step test checks exact array equality between both schedules.
3. Changed the Actor-side missing `model.pipeline.guidance_scale` fallback from `3.5` to `1.0`, matching the documented Actor/rollout split of `1.0`/`3.5`.
4. Replaced the exported `None` fallback with a raise-on-use placeholder when the optional vLLM-Omni rollout backend is unavailable.

Validation:

- Ruff, Ruff format, mypy, generated-config verification, Python compilation, shell syntax, PR-title validation, and `git diff --check`: passed.
- HPSv2 image-conversion CPU tests: `9 passed`.
- Full CPU test suite: `1128 passed, 0 failed` in `49.39s` with exit code `0`.
- Synchronized with upstream `main` at `03655dc` without conflicts (merge commit `ee61dbd`).

The PR body now also includes an MA-5 HPSv2 reward curve over the first 120 equivalent training steps, comparing MindSpeed-MM with the vLLM 0.24.0 and vLLM 0.27.1.dev0 verl-omni runs. 

AI assistance (OpenAI Codex) was used to implement and validate these review updates.
```
```

File: verl_omni/pipelines/flux_dance_grpo/__init__.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    logger.info("FLUX DanceGRPO rollout is unavailable: %s. vLLM-Omni with GPU/NPU support is required.", exc)
+
+    class _UnavailablePipeline:
+        def __getattr__(self, _):
+            raise RuntimeError("FLUX DanceGRPO rollout requires vLLM-Omni with GPU/NPU support")
+
+        def __call__(self, *args, **kwargs):
+            raise RuntimeError("FLUX DanceGRPO rollout requires vLLM-Omni with GPU/NPU support")
+
+    FluxDanceGRPOPipelineWithLogProb = _UnavailablePipeline()
```
Comment (zhtmike) [inline verl_omni/pipelines/flux_dance_grpo/__init__.py]: pls drop this, vllm-omni is not a optional package,

Comment (RhoHwai) [reply verl_omni/pipelines/flux_dance_grpo/__init__.py]: Done — removed the fallback and now import `FluxDanceGRPOPipelineWithLogProb` directly, since `vllm-omni` is a required dependency. Updated in `b91e7dc`. Thanks for the clarification!

**Comment by zhtmike** [comment]
pls make the CI pass locally first, thx.

**Comment by RhoHwai** [comment]
@zhtmike  Thanks for your guidance! All CI checks are now passing.

**Review by zhtmike — APPROVED** [review]

**Comment by zhtmike** [comment]
All CI green. Thanks for nice work!
